addons/binding/org.openhab.binding.plugwise/src/main/java/org/openhab/binding/plugwise/internal/PlugwiseCommunicationHandler.java

```diff
@@ -0,0 +1,591 @@
+            if (queuedMessage.attempts < MAX_RETRIES) {
+                queuedMessage.attempts++;
+
+                Message message = queuedMessage.message;
+
+                String messageHexString = message.toHexString();
+                String packetString = PROTOCOL_HEADER + messageHexString + PROTOCOL_TRAILER;
+                ByteBuffer bytebuffer = ByteBuffer.allocate(packetString.length());
+                bytebuffer.put(packetString.getBytes());
+                bytebuffer.rewind();
+
+                try {
+                    logger.debug("Sending: {} as {}", message, messageHexString);
+                    outputChannel.write(bytebuffer);
+                    sequentialWriteErrors = 0;
+                } catch (IOException e) {
+                    logger.warn("Error writing '{}' to serial port {}: {}", packetString, configuration.getSerialPort(),
+                            e.getMessage());
+                    sequentialWriteErrors++;
+                    return;
+                }
+
+                // Poll the acknowledgement message for at most 1 second, normally it is received within 75ms
+                AcknowledgementMessage ack = acknowledgedQueue.poll(1, TimeUnit.SECONDS);
+                logger.debug("Removing from acknowledgedQueue: {}", ack);
+
+                if (ack == null) {
+                    logger.warn("Error sending: No ACK received after 1 second: {}", packetString);
+                } else if (!ack.isSuccess()) {
+                    if (ack.isError()) {
+                        logger.warn("Error sending: Negative ACK: {}", packetString);
+                    }
+                } else {
+                    // Update the sent message with the new sequence number
+                    message.setSequenceNumber(ack.getSequenceNumber());
+
+                    // Place the sent message in the sent queue
+                    logger.debug("Adding to sentQueue: {}", message);
+                    sentQueueLock.lock();
+                    try {
+                        if (sentQueue.size() == MAX_BUFFER_SIZE) {
+                            // For some reason Plugwise devices, or the Stick, does not send responses to Requests.
+                            // They clog the sent queue. Let's flush some part of the queue
+                            QueuedMessage someMessage = sentQueue.poll();
+                            logger.debug("Flushing from sentQueue: {}", someMessage);
+                        }
+                        sentQueue.put(queuedMessage);
+                    } finally {
+                        sentQueueLock.unlock();
+                    }
+                }
+            } else {
+                // Max attempts reached. We give up, and to a network reset
+                logger.warn("Giving up on Plugwise message after {} attempts: {}", queuedMessage.attempts,
+                        queuedMessage.message);
+            }
+        }
+    }
+
+    private class QueuedMessage {
```

Comment:
I've decomposed the PlugwiseCommunication handler into several objects to reduce the complexity:
* PlugwiseCommunicationContext
* PlugwiseFilteredMessageListener
* PlugwiseFilteredMessageListenerList
* PlugwiseMessagePriority
* PlugwiseMessageProcessor
* PlugwiseMessageSender
* PlugwiseQueuedMessage